docs/architecture/adr-038-state-listening.md

```diff
@@ -269,31 +367,45 @@ func (app *BaseApp) BeginBlock(req abci.RequestBeginBlock) (res abci.ResponseBeg
 
 	...
 
-	// call the hooks with the BeginBlock messages
+	// call the streaming service hook with the BeginBlock messages
 	wg := new(sync.WaitGroup)
-	for _, streamingListener := range app.abciListeners {
-		streamingListener := streamingListener // https://go.dev/doc/faq#closures_and_goroutines
-		if streamingListener.HaltAppOnDeliveryError() {
-			// increment the wait group counter
+	if app.abciListener != nil {
+		if app.stopNodeOnStreamingErr {
 			wg.Add(1)
 			go func() {
-				// decrement the counter when the go routine completes
 				defer wg.Done()
```

Comment:
If you want stopNodeOnStreamingErr to mean that BeginBlock should block until ListenBeginBlock has returned, then there is no reason to run it in a separate goroutine.

```go
...

if app.abciListener != nil {
	if app.stopNodeOnSreamingErr {
		...
		if err := app.abciListener.ListenBeginBlock(...); err != nil {
			...
		}
	} else {
		go func() {
			...
		}()
	}
}

return res
```

Reply:
Very true. Making the change.